content/blog/2022-05-19-CoC/index.md

```diff
@@ -0,0 +1,298 @@
+    forall _:(forall _:P, x), forall _:(forall _:Q, x), x in
+```
+
+Here are a few of the proofs I was able to successfully typecheck.
+```
+(* P /\ Q -> P *)
+Theorem conj_implies_fst: forall P:Type, forall Q:Type, 
+    forall _:((conj P) Q), P.
+Proof. lambda P:Type -> lambda Q:Type -> lambda  p:((conj P) Q) -> 
+    ((p P) (lambda x:P -> lambda _:Q -> x)).
+
+(* P /\ Q -> Q *)
+Theorem conj_implies_snd: forall P:Type, forall Q:Type, 
+    forall _:((conj P) Q), Q.
+Proof. lambda P:Type -> lambda Q:Type -> lambda  p:((conj P) Q) ->
+    ((p Q) (lambda _:P -> lambda x:Q -> x)).
+
+(* P -> Q -> P /\ Q *)
+Theorem p_implies_q_implies_conj: forall P:Type, forall Q:Type, 
+    forall p:P, forall q:Q, ((conj P) Q).
+Proof. lambda P:Type -> lambda Q:Type -> lambda a:P -> lambda b:Q -> 
+    lambda x:Type -> lambda evidence:(forall _:P, forall _:Q, x) -> 
+        ((evidence a) b).
+
+(* P /\ Q -> Q /\ P *)
+Theorem and_commute: forall A:Type, forall B:Type, 
+    forall _:((conj A) B), ((conj B) A).
+Proof. lambda A:Type -> lambda B:Type -> lambda  p:((conj A) B) -> 
+    lambda x:Type -> lambda evidence:(forall _:B, forall _:A, x) ->
+        ((evidence (((conj_implies_snd A) B) p)) (((conj_implies_fst A) B) p)).
+
+(* P -> P \/ Q *)
+Theorem p_implies_sum: forall P:Type, forall Q:Type, forall p:P, ((sum P) Q).
+Proof. lambda P:Type -> lambda Q:Type -> lambda p:P -> lambda x:Type -> 
+    lambda evidence:(forall _:P, x) -> lambda _:(forall _:Q, x) -> (evidence p).
+
+(* P \/ Q -> Q \/ P *)
+Theorem or_commute: forall A:Type, forall B:Type, 
+    forall _:((sum A) B), ((sum B) A).
+Proof. lambda A:Type -> lambda B:Type -> lambda p:((sum A) B) -> 
+    lambda x:Type -> lambda evidence1:(forall _:B, x) -> 
+        lambda evidence2:(forall _:A, x) -> (((p x) evidence2) evidence1).
+```
+
+As a comparison, consider the equivalent lemma and proof for 
+`p_implies_q_implies_conj` in Coq:
+
+```
+Lemma and_intro : forall A B : Prop, A -> B -> A /\ B.
+Proof.
+  intros A B HA HB. split.
+  - apply HA.
+  - apply HB.
+Qed.
+```
+Notice how much cleaner this proof is than my code. It is no wonder that Coq has
+tactics to make it more usable!
+
+My goal was to prove as many one-star exercises in this chapter as possible.
+Out of the eight one-star exercises, these were the ones I was able to prove
+
+```
+Lemma proj2 : forall P Q : Prop,
+  P /\ Q -> Q.
+
+Theorem or_commut : forall P Q : Prop,
+  P \/ Q  -> Q \/ P.
+```
+
+This means I was only able to prove 25% of the one-star exercises. I was not able to prove
+some of the other exercises for a variety of reasons. One exercise, `iff_refl`
+asks you to prove `forall P : Prop, P <-> P` I tried to write this
+proof but for some reason found iff too hard to reason about in CoC. Another exercise, `eqb_neq`,
+asks you to prove `forall x y : nat, x =? y = false <-> x <> y.`. This theorem
+is impossible to prove without defining equality and inequality over the natural 
+numbers, which I did not know how to do. 
```

Comment:
I think you need inductive definitions. That is, you need a classic inductive definition of `nat` like `type nat = zero | succ nat` so you can start saying things about them.